gson/src/main/java/com/google/gson/Gson.java

```diff
@@ -1041,9 +1132,47 @@ private static void assertFullConsumption(Object obj, JsonReader reader) {
   }
 
   /**
-   * Reads the next JSON value from {@code reader} and convert it to an object
+   * Reads the next JSON value from {@code reader} and converts it to an object
+   * of type {@code classOfT}. Returns {@code null}, if the {@code reader} is at EOF.
+   * It is not suitable to use if the specified class is a generic type since it
+   * will not have the generic type information because of the Type Erasure feature of Java.
+   * Therefore, this method should not be used if the desired type is a generic type. Note that
+   * this method works fine if any of the fields of the specified object are generics, just the
+   * object itself should not be a generic type. For the cases when the object is of generic type,
+   * invoke {@link #fromJson(JsonReader, TypeToken)}.
+   *
+   * <p>Unlike the other {@code fromJson} methods, no exception is thrown if the JSON data has
+   * multiple top-level JSON elements, or if there is trailing data.
+   *
+   * <p>The JSON data is parsed in {@linkplain JsonReader#setLenient(boolean) lenient mode},
+   * regardless of the lenient mode setting of the provided reader. The lenient mode setting
+   * of the reader is restored once this method returns.
+   *
+   * @param <T> the type of the desired object
+   * @param reader the reader whose next JSON value should be deserialized
+   * @param classOfT the class of T
+   * @return an object of type T from the JsonReader. Returns {@code null} if {@code reader} is at EOF.
+   * @throws JsonIOException if there was a problem reading from the JsonReader
+   * @throws JsonSyntaxException if json is not a valid representation for an object of type classOfT
+   *
+   * @see #fromJson(Reader, Class)
+   * @see #fromJson(JsonReader, TypeToken)
+   */
+  public <T> T fromJson(JsonReader reader, Class<T> classOfT) throws JsonIOException, JsonSyntaxException {
```

Comment:
Perhaps surprisingly, this is not a source-compatible change. It seems completely logical to add this overload to the other `fromJson(JsonReader...)` ones, but there's an issue. When I ran this PR against all of Google's builds, I encountered [this](https://github.com/googleapis/gax-java/blob/3cba6f7d505a9bb69139239b8f9d155a749e4435/gax-httpjson/src/main/java/com/google/api/gax/httpjson/ApiMessageHttpResponseParser.java#L76):
```java
  private static <ResponseT extends ApiMessage> ApiMessageHttpResponseParser<ResponseT> create(
      final ResponseT responseInstance) {
    final Gson baseGson = new GsonBuilder().create();
    TypeAdapter responseTypeAdapter;
    Gson responseMarshaller = null;

    if (responseInstance != null) {
      responseTypeAdapter =
          new TypeAdapter<ResponseT>() {
            @Override
            public void write(JsonWriter out, ResponseT value) {
              baseGson.toJson(value, responseInstance.getClass(), out);
            }

            @Override
            public ResponseT read(JsonReader in) {
              return baseGson.fromJson(in, responseInstance.getClass());
            }
          };
```

This is not type-safe. The type of `responseInstance.getClass()` is `Class<? extends ApiMessage>`, and `? extends ApiMessage` is not assignable to `ResponseT`. It compiles without this PR because there is no `Class<T>` overload so it gets the type-unsafe `Type` overload.

If the code here instead had `return baseGson.fromJson(in, new TypeToken<ResponseT>() {}.getType())` then it _would_ compile, with or without this PR, and then with this PR the `.getType()` can be removed. It also looks as if this class no longer exists in the current version of `gax-java`.

I found a second problem with code that was calling this method with `Map.class`. The unchecked conversion from raw `Map` to `Map<String, ?>` drew a warning, and the code was being compiled with `-Werror`. The same kind of fix applies, of course.

On the whole, I think we should drop this overload. It's probably better to retain the minor inconsistency of not having it than to expose users to new problems.

Reply:
Thanks for running it against all Google builds.

> If the code here instead had `return baseGson.fromJson(in, new TypeToken<ResponseT>() {}.getType())` then it _would_ compile

But it would not be type safe either, in fact it would even be less type safe than the original code respectively not work properly at all because it would parse the JSON data as the erasure of the type variable (`ApiMessage`). This is why it would have been good to fix #1219, and why I had this originally in #2072 😕. The previous code would not have been type safe though in case `ResponseT` was a parameterized type.

> I found a second problem with code that was calling this method with `Map.class`. The unchecked conversion from raw `Map` to `Map<String, ?>` drew a warning

For what it is worth this would not have been safe either. Even though one might expect that a raw `Map` can only be deserialized to a `Map<String, ...>` due to JSON only permitting strings as member names, Gson actually supports non string values as seen with #1247.

So yes, this will break source compatibility, but it might get users to write safer code, or at least become aware that their code is not safe and then manually suppress the compiler warnings. Therefore, personally I think it might be acceptable to break source compatibility here; but maybe there are cases where the existing code is actually safe and this would create too much trouble...

If you really want, I can remove it though.

(This also goes into the same direction why I mentioned in the description of this pull request to consider deprecating the `fromJson(..., Type)` overloads. That would be rather extreme, but maybe it could be considered at some point in the future.)

Reply:
Well, the title of this PR is _Add Gson.fromJson(..., TypeToken) overloads_, so perhaps we can keep it focused on that. If you want to make a separate PR that adds `fromJson(JsonReader, Class<T>)` then we can debate the source-incompatibility problem there.

The situation is a bit different from #2072, in that already-compiled code will continue to work. It's only new compilations that might fail, and probably only when the code being compiled is doing type-unsafe things. So I think there's a stronger case to be made for the change. Nevertheless, I think it should be separate, if only so I can prepare Google's code-base by fixing the problems I mentioned.

Regarding the type-safety of the first problem, I think the underlying issue is that the method doesn't have enough information to behave correctly. It would need to have something like a `Class<ResponseT>` parameter, and then of course that could be correctly passed to the existing `Type` overload, or to the new `Class<T>` overload. I agree that the new `Class<T>` overload makes it possible to discover this problem, while the new `TypeToken<T>` overload just gives us a _different_ way to sweep the unsoundness under the carpet. I further agree that ideally we would simply fix #1219 to disallow that unsoundness. I haven't changed my mind since #2072, though: I still feel that it's extremely important that people can update their Gson dependency without breaking things that currently work. I'll add that this would be yet _another_ case where a compatibility-level concept might allow us to fix the errors of the past.

Reply:
You are right, I have removed the method again and hope the changes are fine like that.

Sorry for being a bit relentlessly. It is just pretty frustrating and annoying knowing that there are bugs / flaws with Gson and seeing that they cannot be fixed or wasting work on trying to fix them. This is of course not your fault, and I think it is quite valuable that you have a different view of maintaining a library as popular as Gson.